File: AMRAndroid/app/src/main/java/com/amol/amr/navigation/NavigationPlanner.kt
package com.amol.amr.navigation

import kotlin.math.abs

object NavigationPlanner {

    data class PlannerResult(
        val bestDirection: String,
        val costs: Map<String, Float>
    )

    fun planDirection(zones: NavigationZones.ZoneOccupancies): PlannerResult {
        val discountedCenter = (zones.center * 0.8f) - 1e-6f

        val costs = linkedMapOf(
            "FORWARD" to discountedCenter,
            "LEFT" to zones.left,
            "RIGHT" to zones.right
        )

        var minDirection = "FORWARD"
        var minCost = Float.MAX_VALUE

        for ((dir, cost) in costs) {
            if (cost < minCost) {
                minCost = cost
                minDirection = dir
            }
        }

        return PlannerResult(minDirection, costs)
    }

    fun computeSteering(
        zones: NavigationZones.ZoneOccupancies,
        steeringGain: Float = 120.0f,
        maxSteeringAngle: Float = 45.0f,
        deadband: Float = 2.5f
    ): Float {
        // Crowd on right -> positive steering (steer left)
        // Crowd on left -> negative steering (steer right)
        val rawSteering = (zones.right - zones.left) * steeringGain

        if (abs(rawSteering) < deadband) {
            return 0.0f
        }

        return rawSteering.coerceIn(-maxSteeringAngle, maxSteeringAngle)
    }
}

File: src/navigation_planner.py
def navigation_planner(zones):

    left_cost = zones["left"]

    center_cost = zones["center"]

    right_cost = zones["right"]

    # prefer forward motion

    center_cost *= 0.8

    costs = {

        "LEFT": left_cost,

        "FORWARD": center_cost,

        "RIGHT": right_cost
    }

    best_direction = min(costs, key=costs.get)

    return best_direction, costs


# ======================================================
# STEERING COMPUTATION
# ======================================================

def steering_from_zones(zones):

    left = zones["left"]

    center = zones["center"]

    right = zones["right"]

    steering = 0

    # more crowd on left -> steer right
    # more crowd on right -> steer left

    steering += (right - left) * 120

    # limit steering

    steering = max(-45, min(45, steering))

    return steering